tests/e2e/specs/substance.js

```diff
@@ -17,6 +19,66 @@ let qstResponse = {
   ]
 };
 
+describe("The substance form", () => {
+  beforeEach(() => {
+    cy.server();
+    cy.adminLogin();
+    cy.visit("/substance");
+  });
+  it("should validate inputs", () => {
+    cy.get("#casrn").type("not a casrn");
+    cy.get("#substanceType").select("area professor home47");
+    cy.get("#save-substance-btn").click();
+    cy.get("#feedback-casrn").contains(
+      "The proposed CASRN does not conform to the regular expression ^[0-9]{2,7}-[0-9]{2}-[0-9]$"
+    );
+    cy.get("#feedback-substanceType").contains(
+      "The SubstanceType submitted is no longer supported."
+    );
+  });
+  it("should validate nonFieldErrors", () => {
+    let casrn = valid_casrns[Math.floor(Math.random() * valid_casrns.length)];
+    cy.get("#preferredName").type(casrn);
+    cy.get("#casrn").type(casrn);
+    cy.get("#qcLevel").select("QC Level 1");
+    cy.get("#source").select("Source 1");
+    cy.get("#substanceType").select("Substance Type 1");
+    cy.get("#save-substance-btn").click();
+    cy.get("[data-cy=alert-box]").should(
+      "contain",
+      `${casrn} is not unique in ['preferred_name', 'casrn']`
+    );
+    cy.get("#feedback-preferredName").contains("not unique");
+    cy.get("#feedback-casrn").contains("not unique");
+  });
+  // NOT sure that we should do this with it being so difficult to keep the
+  // attrs unique amongst themselves
+  //it("should save valid substance", () => {
+  //  cy.route({
+  //    method: "POST",
+  //    url: "/subtance",
+  //    status: 201,
+  //    response: {}
+  //  }).as("post");
+  //  let casrn = valid_casrns[Math.floor(Math.random() * valid_casrns.length)];
+  //  cy.get("#preferredName").type("preferred substance name");
+  //  cy.get("#casrn").type(casrn);
+  //  cy.get("#qcLevel").select("QC Level 1");
+  //  cy.get("#source").select("Source 1");
+  //  cy.get("#substanceType").select("Substance Type 1");
+  //  cy.get("#save-substance-btn").click();
+  //  //cy.get("[data-cy=alert-box]").should("contain", `created successfully`);
+  //  cy.get("@post").should("have.property", "status", 201);
+  //  cy.get("@post")
+  //    .its("request.body.data.id")
+  //    // This regex accepts only an Oxygen structure
+  //    .should(
+  //      "contain",
+  //      "DTXSID"
+  //    );
+  //})
```

Comment:
every time this test runs it needs a unique `preferredName` and `casrn` cause they have to be unique and the substance gets stored in the DB w/ the test? or is there a way to prevent that?

Reply:
`cy.route({}).as("post");` works as a mock.  Provided the options object is correct `{}`, it will interrupt the browser's natural POST and instead allow you to access that object as "@post".  The response you're currently giving is an empty response (line 61) but if you provided whatever information is required for this functionality (probably nothing), you wont have to worry about the state of the DB at all or any validation.

This will no longer function as a true **e2e** test, and will instead be a **functional** test, but if it's between functional and nothing it'd probably be worth adding just to verify the save button is firing off a post at all.

More info here: https://docs.cypress.io/api/commands/route.html

Reply:
got this working